### PR TITLE
adding nonull environment variable

### PR DESCRIPTION
**Pull request by oddcoder**
This is supposed to help delaying PC divergence to null by preventing all null accesses in PC and memory
I am not sure if this will be enough.

### CODE REVIEW

File: libr/anal/esil.c [hunk 1]
```diff
@@ -235,6 +235,22 @@ static int internal_esil_mem_write(RAnalEsil *esil, ut64 addr, const ut8 *buf, i
 	return ret;
 }
 
+static int internal_esil_mem_write_no_null(RAnalEsil *esil, ut64 addr, const ut8 *buf, int len) {
+	int ret;
+	if (!esil || !esil->anal || !esil->anal->iob.io || !addr)
```
Comment (radare) [inline libr/anal/esil.c]: missing braces

File: libr/anal/esil.c [hunk 2]
```diff
@@ -235,6 +235,22 @@ static int internal_esil_mem_write(RAnalEsil *esil, ut64 addr, const ut8 *buf, i
...
+		return 0;
+	if (esil->nowrite)
```
Comment (radare) [inline libr/anal/esil.c]: missing braces

File: libr/anal/esil.c [hunk 3]
```diff
@@ -272,6 +288,18 @@ static int internal_esil_reg_write(RAnalEsil *esil, const char *regname, ut64 nu
...
+	RRegItem *reg = r_reg_get (esil->anal->reg, regname, -1);
+	const char *pc = r_reg_get_name (esil->anal->reg, R_REG_NAME_PC);
+	const char *sp = r_reg_get_name (esil->anal->reg, R_REG_NAME_SP);
+	const char *bp = r_reg_get_name (esil->anal->reg, R_REG_NAME_BP);
+	if (reg && reg->name && ((strcmp (reg->name , pc) && strcmp (reg->name, sp) && strcmp(reg->name, bp)) || num)) { //I trust k-maps
+		r_reg_set_value (esil->anal->reg, reg, num);
+		return true;
+	}
+	return false;
+
```
Comment (radare) [inline libr/anal/esil.c]: empty line

File: libr/anal/esil.c [hunk 4]
```diff
@@ -272,6 +288,18 @@ static int internal_esil_reg_write(RAnalEsil *esil, const char *regname, ut64 nu
 	}
 	return false;
 }
+static int internal_esil_reg_write_no_null (RAnalEsil *esil, const char *regname, ut64 num) {
+	RRegItem *reg = r_reg_get (esil->anal->reg, regname, -1);
```
Comment (radare) [inline libr/anal/esil.c]: check if esil, esil->anal and esil->anal->reg is null at least

File: libr/anal/esil.c [hunk 5]
```diff
@@ -272,6 +288,18 @@ static int internal_esil_reg_write(RAnalEsil *esil, const char *regname, ut64 nu
...
+	const char *pc = r_reg_get_name (esil->anal->reg, R_REG_NAME_PC);
+	const char *sp = r_reg_get_name (esil->anal->reg, R_REG_NAME_SP);
+	const char *bp = r_reg_get_name (esil->anal->reg, R_REG_NAME_BP);
+	if (reg && reg->name && ((strcmp (reg->name , pc) && strcmp (reg->name, sp) && strcmp(reg->name, bp)) || num)) { //I trust k-maps
+		r_reg_set_value (esil->anal->reg, reg, num);
```
Comment (radare) [inline libr/anal/esil.c]: if num != NULL this line will be executed with reg = NULL

**Comment by radare** [comment]
there are many tests crashing, pls fix those crashes

File: libr/anal/esil.c (hunk 5 above)
Comment (oddcoder) [inline libr/anal/esil.c]: yes very very specific set of registers like the the PC, BP and SP

Comment (oddcoder) [inline libr/anal/esil.c]: if any of those 3 registers contain zero then something should be wrong

**Comment by oddcoder** [comment]
tests are updated here https://github.com/radare/radare2-regressions/pull/495

**Comment by oddcoder** [comment]
actually another issue arose from preventing null accesses is that they are no longer logged into `aek *` 
which doesn't solve the problem this pr was made for.
I should still find out how to log the memory accesses into `aek *` without executing them

**Comment by oddcoder** [comment]
updated and ready for merge I guess

**Comment by radare** [comment]
Maybe disable it by default because there are many tests that are not passing

**Comment by oddcoder** [comment]
actually there was just 8 broken test cases all was something like mov eax, [0] stuff thought I disabled it by default

**Comment by radare** [comment]
lets merge it for now. we can check those issues later.

**Pull request merged**